cpp/src/lookup/lookup_src_dst_impl.cuh

```diff
@@ -0,0 +1,1058 @@
+      std::vector<edge_t> h_type_freqs_segment(nr_uniqe_types_segment);
+
+      raft::update_host(h_unique_types_segment.data(),
+                        unique_types_segment.data(),
+                        unique_types_segment.size(),
+                        handle.get_stream());
+      raft::update_host(h_type_freqs_segment.data(),
+                        type_freqs_segment.data(),
+                        type_freqs_segment.size(),
+                        handle.get_stream());
+
+      if constexpr (multi_gpu) {
+        h_gpu_ids_partition.insert(h_gpu_ids_partition.end(), nr_uniqe_types_segment, gpu_id);
+      }
+
+      h_types_partition.insert(
+        h_types_partition.end(), h_unique_types_segment.begin(), h_unique_types_segment.end());
+      h_type_freqs_partition.insert(
+        h_type_freqs_partition.end(), h_type_freqs_segment.begin(), h_type_freqs_segment.end());
+    }
+
+    [[maybe_unused]] rmm::device_uvector<int> gpu_ids_partition(h_gpu_ids_partition.size(),
+                                                                handle.get_stream());
+    rmm::device_uvector<edge_type_t> types_partition(h_types_partition.size(), handle.get_stream());
+    rmm::device_uvector<edge_t> type_freqs_partition(h_type_freqs_partition.size(),
+                                                     handle.get_stream());
+
+    if constexpr (multi_gpu) {
+      raft::update_device(gpu_ids_partition.data(),
+                          h_gpu_ids_partition.data(),
+                          h_gpu_ids_partition.size(),
+                          handle.get_stream());
+    }
+    raft::update_device(types_partition.data(),
+                        h_types_partition.data(),
+                        h_types_partition.size(),
+                        handle.get_stream());
+    raft::update_device(type_freqs_partition.data(),
+                        h_type_freqs_partition.data(),
+                        h_type_freqs_partition.size(),
+                        handle.get_stream());
+
+    if constexpr (multi_gpu) {
+      // Shuffle to the proper GPUs
+      std::forward_as_tuple(std::tie(gpu_ids_partition, types_partition, type_freqs_partition),
+                            std::ignore) =
+        cugraph::groupby_gpu_id_and_shuffle_values(
+          handle.get_comms(),
+          thrust::make_zip_iterator(thrust::make_tuple(
+            gpu_ids_partition.begin(), types_partition.begin(), type_freqs_partition.begin())),
+          thrust::make_zip_iterator(thrust::make_tuple(
+            gpu_ids_partition.end(), types_partition.end(), type_freqs_partition.end())),
+          [] __device__(auto val) { return thrust::get<0>(val); },
+          handle.get_stream());
+
+      thrust::for_each(
+        handle.get_thrust_policy(),
+        gpu_ids_partition.begin(),
+        gpu_ids_partition.end(),
+        [rank = handle.get_comms().get_rank()] __device__(auto val) { assert(val == rank); });
```

Comment:
Better delete this once this code gets validated.

Reply:
The part of the code has been removed now.